source-code/core2/src/fullExample.ts

```diff
@@ -0,0 +1,105 @@
+import { createApp } from "./app/createApp.js"
+import type { InlangConfig } from "./config/schema.js"
+import type {  MessageLintRule } from "./lint/api.js"
+import type { Plugin_Proposal_2 } from "./plugin/api.js"
+
+// --------------------- LINT RULE ---------------------
+
+const missingMessage: MessageLintRule = {
+	id: "inlang.missingMessage",
+	displayName: {
+		en: "Missing message",
+		de: "Fehlende Nachricht",
+	},
+	defaultLevel: "error",
+	type: "Message",
+	message: ({ message, inlang }) => {
```

Comment:
I would keep the API, that we already have currently.
Here you forgot to pass the lint level and the id of the lint rule. Sure this can be altered afterwards, but I would keep the `.report()` API.

Reply:
The report api makes the implementation vastly more complex. 

Simply returning a LintReport object is much simpler both internally and externally (explaining the system). Hence, I vote against `report`.

Reply:
```ts
const missingMessage: MessageLintRule = {
	id: "inlang.missingMessage",
	displayName: {
		en: "Missing message",
		de: "Fehlende Nachricht",
	},
	defaultLevel: "error",
	type: "Message",
	message: ({ message, inlang, report }) => {
		if (message.languageTag !== inlang.config.sourceLanguageTag) {
			return
		}
		// TODO pain to make reports typesafe...
		// - MessageLintReport has a lot of duplicate information
		for (const languageTag of inlang.config.languageTags) {
			const translation = inlang.messages.query.get({ where: { id: message.id, languageTag } })
			if (!translation) {
				report({
					languageTag,
					messageId: message.id,
					content: `Message "${message.id}" is missing in language tag "${languageTag}".`,
				})
			}
		}
	},
}
```

no need to return an array, just call the `report` function.

Reply:
PR is out https://github.com/inlang/inlang/pull/1154#issue-1815593130. 

Closing this discussion in favor of the PR>